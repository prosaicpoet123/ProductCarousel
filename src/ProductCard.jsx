import React from "react";

export default ({ item }) => {
  return (
    <div className="ProductCard">
      <img src={item.image} alt={item.title} />
      <h2>{item.title}</h2>
    </div>
  );
};
