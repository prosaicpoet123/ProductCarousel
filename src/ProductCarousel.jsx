import React, { useState, useLayoutEffect } from "react";
import ProductCard from "./ProductCard";

function useWindowSize() {
  const [size, setSize] = useState([0, 0]);
  useLayoutEffect(() => {
    function updateSize() {
      setSize([window.innerWidth]);
    }
    window.addEventListener('resize', updateSize);
    updateSize();
    return () => window.removeEventListener('resize', updateSize);
  }, []);
  return size;
}

const ProductCarousel = ({ list, itemsToShow = 5, skipBy = 1 }) => {
  const [carousel, updateCarousel] = useState({
    width: 150,
    translate: 0,
    position: 0
  });
  const [width] = useWindowSize();
  let productlist = React.createRef();
  useLayoutEffect(() => {
    const { width } = productlist.current.getBoundingClientRect();
    updateCarousel({ width: width / itemsToShow, translate: 0, position: 0 });
  }, [width]);
  if (!list) return null;
  const divStyle = {
    width: `${list.length * carousel.width}px`,
    transform: `translate3d(${carousel.translate}px, 0px, 0px)`,
    transition: "transform 0.5s ease-in-out"
  };
  const itemStyle = {
    width: `${carousel.width}px`
  };
  const handleOnClick = direction => {
    const newTranslation =
      direction === "next"
        ? carousel.translate - carousel.width * skipBy
        : carousel.translate + carousel.width * skipBy;
    const newPosition =
      direction === "next"
        ? carousel.position + 1 * skipBy
        : carousel.position - 1 * skipBy;
    updateCarousel({
      width: carousel.width,
      translate: newTranslation,
      position: newPosition
    });
  };
  return (
    <div className="ProductCarousel">
      <div className="CarouselList" ref={productlist}>
        <div className="CarouselSlider" style={divStyle}>
          {list.map((item, i) => {
            return (
              <div className="CarouselItem" style={itemStyle}>
                <ProductCard item={item} key={i} />
              </div>
            );
          })}
        </div>
        <div className="CarouselUI">
          <button
            className="backBtn"
            onClick={e => handleOnClick("back")}
            disabled={carousel.position === 0}
          >
            Back
          </button>
          <button
            className="nextBtn"
            onClick={e => handleOnClick("next")}
            disabled={carousel.position === list.length - itemsToShow}
          >
            Next
          </button>
        </div>
      </div>
    </div>
  );
};

export default ProductCarousel;

//https://stackoverflow.com/questions/19014250/rerender-view-on-browser-resize-with-react
