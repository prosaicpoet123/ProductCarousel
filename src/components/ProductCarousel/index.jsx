import React, { useState, useLayoutEffect } from "react";
import ProductCard from "../ProductCard";

function useWindowSize() {
  const [size, setSize] = useState([0, 0]);
  useLayoutEffect(() => {
    function updateSize() {
      setSize([window.innerWidth]);
    }
    window.addEventListener("resize", updateSize);
    updateSize();
    return () => window.removeEventListener("resize", updateSize);
  }, []);
  return size;
}

function handleResponsiveData(data, windowWidth) {
  return data.find(item => windowWidth < item.breakpoint);
}

function getNewPositions(direction, carousel) {
  const position =
    direction === "next"
      ? carousel.position + 1 * carousel.itemsToScroll
      : carousel.position - 1 * carousel.itemsToScroll;
  const translate =
    direction === "next"
      ? carousel.translate - carousel.width * carousel.itemsToScroll
      : carousel.translate + carousel.width * carousel.itemsToScroll;
  return { position, translate };
}

const ProductCarousel = ({
  list,
  itemsToShow = 5,
  itemsToScroll = 1,
  speed = 500,
  responsive = []
}) => {
  const [carousel, updateCarousel] = useState({
    width: 150,
    translate: 0,
    position: 0,
    itemsToScroll,
    itemsToShow
  });
  const [windowWidth] = useWindowSize();
  const productlist = React.createRef();
  useLayoutEffect(() => {
    const { width } = productlist.current.getBoundingClientRect();
    const responsiveData = handleResponsiveData(responsive, windowWidth);
    updateCarousel({
      position: 0,
      translate: 0,
      itemsToShow: responsiveData
        ? responsiveData.settings.itemsToShow
        : itemsToShow,
      itemsToScroll: responsiveData
        ? responsiveData.settings.itemsToScroll
        : itemsToScroll,
      width:
        width /
        (responsiveData ? responsiveData.settings.itemsToShow : itemsToShow)
    });
  }, [windowWidth]);
  if (!list || !list.length) return null;
  const handleOnClick = direction => {
    const { position, translate } = getNewPositions(direction, carousel);
    console.log(position);
    console.log(list.length - carousel.itemsToShow);
    updateCarousel({
      ...carousel,
      width: carousel.width,
      translate,
      position
    });
  };
  return (
    <div className="ProductCarousel">
      <div className="CarouselList">
        <button
          className="backBtn"
          onClick={e => handleOnClick("back")}
          disabled={carousel.position === 0}
        >
          Back
        </button>
        <div className="CarouselSlider" ref={productlist}>
          <div
            className="CarouselTrack"
            style={{
              width: `${list.length * carousel.width}px`,
              transform: `translate3d(${carousel.translate}px, 0px, 0px)`,
              transition: `transform ${speed}ms ease-in-out`
            }}
          >
            {list.map((item, i) => {
              return (
                <div
                  className="CarouselItem"
                  key={i}
                  style={{ width: `${carousel.width}px` }}
                >
                  <ProductCard item={item} />
                </div>
              );
            })}
          </div>
        </div>
        <button
          className="nextBtn"
          onClick={e => handleOnClick("next")}
          disabled={carousel.position === list.length - carousel.itemsToShow}
        >
          Next
        </button>
      </div>
    </div>
  );
};

export default ProductCarousel;
