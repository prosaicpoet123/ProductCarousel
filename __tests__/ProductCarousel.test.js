import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import ProductCarousel from "../src/components/ProductCarousel";
import ProductCard from "../src/components/ProductCard";

const productList = require("../src/data/productList.json");

describe("ProductCarousel Snapshot Tests", () => {
  let productCarouselElement;
  beforeEach(() => {
    productCarouselElement = renderer.create(
      <ProductCarousel>
        {productList.map(item => (
          <ProductCard item={item} />
        ))}
      </ProductCarousel>
    );
  });
  it("Renders the ProductCarousel Correctly", () => {
    let tree = productCarouselElement.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("ProductCarousel shallow rendering tests", () => {
  it("Should display the correct number of items", () => {
    let carousel = mount(
      <ProductCarousel>
        {productList.map(item => (
          <ProductCard item={item} />
        ))}
      </ProductCarousel>
    );
    expect(carousel.find(".CarouselItem").length).toBe(10);
  });
});
