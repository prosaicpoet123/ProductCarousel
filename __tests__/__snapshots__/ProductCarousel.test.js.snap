// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCarousel Snapshot Tests Renders the ProductCarousel Correctly 1`] = `
<div
  className="ProductCarousel"
>
  <div
    className="CarouselList"
  >
    <button
      className="backBtn"
      disabled={true}
      onClick={[Function]}
    >
      Back
    </button>
    <div
      className="CarouselSlider"
    >
      <div
        className="CarouselTrack"
        style={
          Object {
            "transform": "translate3d(0px, 0px, 0px)",
            "transition": "transform 500ms ease-in-out",
            "width": "NaNpx",
          }
        }
      >
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 1"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 2"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 3"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 4"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 5"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 6"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 7"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 8"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 9"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
        <div
          className="CarouselItem"
          style={
            Object {
              "width": "NaNpx",
            }
          }
        >
          <div
            className="ProductCard"
          >
            <img
              alt="Product 10"
              src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQoW07vtqK4oFr1gVi7dULROdDjAFU9r6FZwlFo7CM3pYfEDV9-"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      className="nextBtn"
      disabled={false}
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;